libraries/AP_InertialSensor/BatchSampler.cpp

```diff
@@ -0,0 +1,198 @@
+#include "AP_InertialSensor.h"
+#include <GCS_MAVLink/GCS.h>
+
+// Class level parameters
+const AP_Param::GroupInfo AP_InertialSensor::BatchSampler::var_info[] = {
+    // @Param: BAT_CNT
+    // @DisplayName: sample count per batch
+    // @Description: Number of samples to take when logging streams of IMU sensor readings.  Will be rounded down to a multiple of 32.
+    // @User: Advanced
+    // @Increment: 32
+    AP_GROUPINFO("BAT_CNT",  1, AP_InertialSensor::BatchSampler, _required_count,   1024),
+
+    // @Param: BAT_MASK
+    // @DisplayName: Sensor Bitmask
+    // @Description: Bitmap of which IMUs to log batch data for
+    // @User: Advanced
+    // @Values: 0:None,1:First IMU,255:All
+    // @Bitmask: 0:IMU1,1:IMU2,2:IMU3
+    AP_GROUPINFO("BAT_MASK",  2, AP_InertialSensor::BatchSampler, _sensor_mask,   DEFAULT_IMU_LOG_BAT_MASK),
+
+    AP_GROUPEND
+};
+
+
+extern const AP_HAL::HAL& hal;
+void AP_InertialSensor::BatchSampler::init()
+{
+    if (_sensor_mask == 0) {
+        return;
+    }
+    if (_required_count <= 0) {
+        return;
+    }
+
+    _required_count -= _required_count % 32; // round down to nearest multiple of 32
+
+    const uint32_t total_allocation = 3*_required_count*sizeof(uint16_t);
+    gcs().send_text(MAV_SEVERITY_WARNING, "INS: alloc %u bytes for ISB (free=%u)", total_allocation, hal.util->available_memory());
+
+    data_x = (int16_t*)calloc(_required_count, sizeof(int16_t));
+    data_y = (int16_t*)calloc(_required_count, sizeof(int16_t));
+    data_z = (int16_t*)calloc(_required_count, sizeof(int16_t));
+    if (data_x == nullptr || data_y == nullptr || data_z == nullptr) {
+        free(data_x);
+        free(data_y);
+        free(data_z);
+        data_x = nullptr;
+        data_y = nullptr;
+        data_z = nullptr;
+        gcs().send_text(MAV_SEVERITY_WARNING, "Failed to allocate %u bytes for IMU batch sampling", total_allocation);
+        return;
+    }
+
+    initialised = true;
+}
+
+void AP_InertialSensor::BatchSampler::periodic()
+{
+    if (_sensor_mask == 0) {
+        return;
+    }
+    push_data_to_log();
+}
+
+void AP_InertialSensor::BatchSampler::rotate_to_next_sensor()
+{
+    const uint8_t old_instance = instance;
+    do {
+        instance++;
+        switch(type) {
+        case IMU_SENSOR_TYPE_ACCEL:
+            if (instance >= _imu._accel_count) {
+                type = IMU_SENSOR_TYPE_GYRO;
+                instance = 0;
+                multiplier = multiplier_accel;
+            }
+            break;
+        case IMU_SENSOR_TYPE_GYRO:
+            if (instance >= _imu._gyro_count) {
+                type = IMU_SENSOR_TYPE_ACCEL;
+                instance = 0;
+                multiplier = multiplier_gyro;
```

Comment:
this should be multipler_accel
I think this should the code should be restructured a bit to make it clearer. Maybe a first switch to change the type, then a 2nd switch to setup the new multuplier

Reply:
Will fix.  I've just factored out the setting of multiplier.  It doesn't help much.